pmd-modelica/etc/grammar/Modelica.jjt

```diff
@@ -0,0 +1,952 @@
+        LogicalExpression()
+        (
+            <COLON>
+            LogicalExpression()
+        )?
+    )?
+}
+
+void LogicalExpression(): {}
+{
+    LogicalTerm()
+    (
+        <OR>
+        LogicalTerm()
+    )*
+}
+
+void LogicalTerm(): {}
+{
+    LogicalFactor()
+    (
+        <AND>
+        LogicalFactor()
+    )*
+}
+
+void LogicalFactor(): {}
+{
+    (<NOT>)?
+    Relation()
+}
+
+void Relation(): {}
+{
+    ArithmeticExpression()
+    (
+        RelOp()
+        ArithmeticExpression()
+    )?
+}
+
+void RelOp(): {}
+{
+  <LT> | <LE> | <GT> | <GE> | <EQ> | <NEQ>
+}
+
+void ArithmeticExpression(): {}
+{
+    (AddOp())?
+    Term()
+    (
+        AddOp()
+        Term()
+    )*
+}
+
+void AddOp(): {}
+{
+    <PLUS> | <MINUS> | <DOT_PLUS> | <DOT_MINUS>
+}
+
+void Term(): {}
+{
+    Factor()
+    (
+        MulOp()
+        Factor()
+    )*
+}
+
+void MulOp(): {}
+{
+    <ASTERISK> | <SLASH> | <DOT_ASTERISK> | <DOT_SLASH>
+}
+
+void Factor(): {}
+{
+    Primary()
+    (
+        (<HAT> | <DOT_HAT>)
+        Primary()
+    )?
+}
+
+void Primary(): {}
+{
+      <UNSIGNED_NUMBER>
+    | <STRING>
+    | <FALSE>
+    | <TRUE>
+    | ((Name() | <DER> | <INITIAL>) FunctionCallArgs())
+    | ComponentReference()
+    | (<LPAREN> OutputExpressionList() <RPAREN>)
+    | (<LSQUARE> ExpressionList() (<SC> ExpressionList())* <RSQUARE>)
+    | (<LCURL> FunctionArguments() <RCURL>)
+    | <END>
+}
+
+void Name(): {}
+{
+    (<DOT>)?
+    <IDENT>
+    (
+        <DOT>
+        <IDENT>
+    )*
+}
+
+void ComponentReference(): {}
+{
+    (<DOT>)?
+    <IDENT>
+    (ArraySubscripts())?
+    (
+        <DOT>
+        <IDENT>
+        (ArraySubscripts())?
+    )*
+}
+
+void FunctionCallArgs(): {}
+{
+    <LPAREN>
+    (FunctionArguments())?
+    <RPAREN>
+}
+
+void FunctionArguments(): {}
+{
+      (FunctionArgument() (<COMMA> FunctionArguments() | (<FOR> ForIndices()))?)
+    | NamedArguments()
+}
+
+void NamedArguments(): {}
+{
+    NamedArgument()
+    (
+        <COMMA>
+        NamedArguments()
+    )?
+}
+
+void NamedArgument(): {}
+{
+    <IDENT>
+    <EQ_SIGN>
+    FunctionArgument()
+}
+
+void FunctionArgument(): {}
+{
+    (
+        <FUNCTION>
+        Name()
+        <LPAREN>
+        (NamedArguments())?
+        <RPAREN>
+    ) | (
+        Expression()
+    )
+}
+
+void OutputExpressionList(): {}
+{
+    (Expression())?
+    (
+        <COMMA>
+        (Expression())?
+    )*
+}
+
+void ExpressionList(): {}
+{
+    Expression()
+    (
+        <COMMA>
+        Expression()
+    )*
+}
+
+void ArraySubscripts(): {}
+{
+    <LSQUARE>
+    Subscript()
+    (
+        <COMMA>
+        Subscript()
+    )*
+    <RSQUARE>
+}
+
+void Subscript(): {}
+{
+      <COLON>
+    | Expression()
+}
+
+void Comment(): {}
+{
+    StringComment()
+    (Annotation())?
+}
+
+void StringComment(): {}
+{
+    (
+        <STRING>
+        (<PLUS> <STRING>)*
+    )?
+}
+
+void Annotation(): {}
+{
+    <ANNOTATION>
+    ClassModification()
+}
+
+TOKEN: {
```

Comment:
It's good practice to declare all tokens together at the top of the file. Otherwise JavaCC may generate implicit tokens at the first point of use in some specific cases...

Also, related to the rest of the grammar: if the lexer is case-sensitive (which I assume yours is), you don't have to refer to tokens as `<THEN>`, you can just write `"then"`, which improves readability.

Reply:
> Also, related to the rest of the grammar: if the lexer is case-sensitive (which I assume yours is), you don't have to refer to tokens as <THEN>, you can just write "then", which improves readability.

Not sure it is a big problem, because, on the other hand, they are easily distinguishable even without the syntax highlighting, so it looks like both approaches have their pros and cons. I see both styles from language to language in the PMD repository. I don't argue against being requested to change this, though. (The other question, maybe I misunderstood some JJTree syntax, `[ "a" ]` expressions for example, are somewhere referenced as equivalents of `("a")?`, if I get it right...)

Reply:
> It's good practice to declare all tokens together at the top of the file. Otherwise JavaCC may generate implicit tokens at the first point of use in some specific cases...

I'm not so familiar with JavaCC yet, but if this can eliminate some obscure debugging, then I should definitely fix this. Maybe after some other review comments.

Reply:
> it looks like both approaches have their pros and cons.

Yes, ultimately that's up to your preference. I usually use syntax highlighting and prefer the less ceremonial form with quotes, but whatever. I don't expect you to change it

>` [ "a" ]` expressions for example, are somewhere referenced as equivalents of `("a")?`

Yes the form with square brackets is just sugar, sometimes it's more readable... I use those forms inconsistently myself tbh

Reply:
I just want to say it looks like regular expression token, e.g., in `cpp.jj`:
```
<  OCTALINT : "0" (["'", "0"-"7"])* >
```
but has different meaning. So it was kind of confusing for me when I read _different_ tutorials :)

Reply:
Oh yeah right, the Javacc language has two sub-languages for regular expressions and BNF expansions, differing only in subtle ways. In a regular expression context,  `[...]` means a character class, while in BNF it's like `(...)?`. Regular expressions more complicated than just a string literal are all enclosed in `< ... >`, so you can identify the context like that. 

Side note: I think it's probably a very old mistake from the JavaCC guys to require string literals in character classes. They could have required character literals, and this would have been much less ambiguous (and for them, easier to parse).

Reply:
> It's good practice to declare all tokens together at the top of the file.

Fixed